File: Src/ButterflyNet.Parser/Strategies/DefinitionStrategy.cs
using ButterflyNet.Parser.Satisfiers;
using ButterflyNet.Parser.Scopes;

namespace ButterflyNet.Parser.Strategies {
	public class DefinitionStrategy : ScopeDrivenStrategy {
		public DefinitionStrategy() {
			AddSatisfier<StartOfLineSatisfier>();
			AddSatisfier(new InScopeStackSatisfier(ScopeTypeCache.DefinitionList));
			AddSatisfier(new ExactCharMatchSatisfier(":"));
		}

		protected override void DoExecute(ParseContext context) {
			OpenScope(new DefinitionScope(), context);
		}

		private class TermMustBePreviousNodeSatisfier : ISatisfier {
			public bool IsSatisfiedBy(ParseContext context) {
				var previousNode = context.ScopeTree.GetMostRecentNode(depth: context.Scopes.Count);
				return previousNode != null && previousNode.Scope.GetType() == ScopeTypeCache.DefinitionTerm;
			}
		}
	}
}

File: src/js/src/strategies/DefinitionStrategy.js
function DefinitionStrategy() {
	DefinitionStrategy.$parent.call(this);
	
	this.addSatisfier(new StartOfLineSatisfier());
	this.addSatisfier(new InScopeStackSatisfier(ScopeTypeCache.definitionList));
	this.setAsTokenTransformer(":");
	this.addSatisfier(function(context) {
		var previousNode = context.scopeTree.getMostRecentNode(context.scopes.length);
		return previousNode && previousNode.scope.type === ScopeTypeCache.definitionTerm;
	});
	
	this.doExecute = function(context) {
		this.openScope(new DefinitionScope(), context);
	};
}

extend(ScopeDrivenStrategy, DefinitionStrategy);